### ISSUE TITLE
Implementing skill selector

### ISSUE THREAD

**Issue by tfederico** [xbpeng/DeepMimic#136]
Hello,

I am trying to implement the skill selector for switching between multiple clips.

I came up with two different options on how to implement it:

1) sampling $g_{t}$ when calling `reset()`, using the corresponding clip $i$ and the goal $g_{t}$ as an additional input to the policy, calculating the reward only for the clip $i$
2) calculating the reward for all the clips as during the multiclip training and then selecting the reward for clip $i$ using $\prod_{j=1}^{k} g_{t}^{j} * r_{t}^{j}$ where $g_{t}^{j} = 1$ if $i=j$ and $g_{t}^{j} = 0$ otherwise.

The former looks more similar to what you described in the paper, but it eludes me how it learn to transit between different motions if the change of clip happens when you reset the character.

**Comment by tfederico** [comment]
Apologies, yesterday I wrote <img src="http://latex.codecogs.com/svg.latex?\inline&space;\prod_{j=1}^{k}&space;g_{t}^{j}&space;r_{t}^{j}" title="http://latex.codecogs.com/svg.latex?\inline \prod_{j=1}^{k} g_{t}^{j} r_{t}^{j}" /> but what I really meant was <img src="http://latex.codecogs.com/svg.latex?\underset{j&space;=&space;1..k}{max}&space;g_{t}^{j}&space;r_{t}^{j}" title="http://latex.codecogs.com/svg.latex?\underset{j = 1..k}{max} g_{t}^{j} r_{t}^{j}" />

**Comment by xbpeng** [comment]
Option 1 sounds right. Providing a goal input to the policy that indicates which clip it should imitate, and calculate the reward with respect to the selected clip. To train the policy to transition between different clips, you can change the goal every now and then during an episode. This will force the policy to learn how to transition between the different clips.

**Comment by tfederico** [comment]
Great, thanks! Did you find any optimal value for the time to change between different clips? 

Moreover, I realized that to switch between different clips I need to translate the root position of the new clip based on the current position of the character. Is there any other adjustment that I need to take care of when I switch to another clip?

**Comment by tfederico** [comment]
In addition, did you fine any limit to the number of similar clips that you can select?

**Comment by xbpeng** [comment]
We hold each clip between 5-10s before switching to a new one, and yes, when switching to a new clip, we also synchronize the root position of the character and reference motion.

It can be tricky to train a single policy to imitate a lot of different motions. That tends to lead to the policy not imitating any of the motions very well. The max number of motions depend on how different the motions are. If all of your clips are walking motions, then we can imitate more of them. The max we've tried was around 10. But if the motions are more diverse, like the different kinds of flips shown in the video, then the max we were able to learn was 4 at a time.

**Comment by tfederico** [comment]
Did you also synchronize the root rotation? i am trying to figure out how to synchronize it but I can't figure out how.
What I did for the root position was:
1) get the current root position from the kinematic character
2) assign it as the root position of the new clip at time 0
3) calculate the root translation in the new clip from t to t+1
4) add the translation to the current root position

I was thinking of doing something similar for the root rotation but I can't figure how to calculate the difference between the quaternions. Any help would be very much appreciated, thanks!

**Comment by xbpeng** [comment]
We do the synchronization a little bit differently. When switching to a new clip, we apply an offset so that the root position is at the same position as the root of the simulated character. Since over the course of the motion, the sim character might drift away from the kinematic character, we just resync the positions whenever the clip changes. Note that we only synchronize the x and z positions of the root, the y position stay the same as the original clip.

The root rotation is similar, we sync rotation to the simulated character's root rotation, instead of the kinematic character's. But one detail here is that we only synchronize the heading direction, not the full rotation. The heading can be calculated with this:
https://github.com/xbpeng/DeepMimic/blob/687945c8abde39fd24e1d175cc155f7cb8845729/DeepMimicCore/anim/KinTree.cpp#L1621
We then just apply a rotation offset to the new clip so that the heading will start out being the same as the simulated character's heading.

**Comment by tfederico** [comment]
So given the heading of the simulated character and the one of the clip, how do you calculate the rotation offset to apply to the character root rotation?Do you take the difference between the two heading and somehow convert it into a quaternion? Apologies for the basic questions but I am not really an expert when it comes to kinematics

**Comment by xbpeng** [comment]
Yes, that's right. To calculate the rotation offset, you can compute the heading of the two characters, that will give you a scalar rotation around the y-axis. You can then compute the difference and covert that to a quaternion. Then rotate the kinematic character's root by that delta rotation to align the heading of the two characters.

**Comment by tfederico** [comment]
so what I did was:

1. get root orientation of simulated character and the root rotation of the new clip
2. calculate root direction based on the orientation
3. calculate the two headings using `atan2`
4. calculate the heading offset as the difference between the simulation heading and the clip heading
5. converting the heading offset (euler) to a quaternion
6. apply the delta rotation to the clip, so that the root rotation is alligned to the simulated character

In this way, the new clip root rotation should be aligned to the simulated character one. However, it appears that sometimes there is a 180° mismatch around the Y axis.
Do you have any idea what might be causing this?

**Comment by tfederico** [comment]
@xbpeng is there any module operation (e.g., heading offset modulo 2*pi) or offset that need to be computed?
